clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -1468,6 +1467,145 @@ test_replay_skip()
+        sched1.emplace_back(TID_BASE, 110, CPU1, 7);
+        sched1.emplace_back(TID_BASE + 2, 112, CPU1, 7);
+        std::ostringstream cpu0_string;
+        cpu0_string << CPU0;
+        std::ostringstream cpu1_string;
+        cpu1_string << CPU1;
+        zipfile_ostream_t outfile(cpu_fname);
+        std::string err = outfile.open_new_component(cpu0_string.str());
+        assert(err.empty());
+        if (!outfile.write(reinterpret_cast<char *>(sched0.data()),
+                           sched0.size() * sizeof(sched0[0])))
+            assert(false);
+        err = outfile.open_new_component(cpu1_string.str());
+        assert(err.empty());
+        if (!outfile.write(reinterpret_cast<char *>(sched1.data()),
+                           sched1.size() * sizeof(sched1[0])))
+            assert(false);
+    }
+
+    // Replay the recorded schedule.
+    std::vector<scheduler_t::input_workload_t> sched_inputs;
+    for (int i = 0; i < NUM_INPUTS; i++) {
+        memref_tid_t tid = TID_BASE + i;
+        std::vector<scheduler_t::input_reader_t> readers;
+        readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(inputs[i])),
+                             std::unique_ptr<mock_reader_t>(new mock_reader_t()), tid);
+        sched_inputs.emplace_back(std::move(readers));
+    }
+    scheduler_t::scheduler_options_t sched_ops(scheduler_t::MAP_TO_RECORDED_OUTPUT,
+                                               scheduler_t::DEPENDENCY_TIMESTAMPS,
+                                               scheduler_t::SCHEDULER_DEFAULTS,
+                                               /*verbosity=*/4);
+    zipfile_istream_t infile(cpu_fname);
+    sched_ops.replay_as_traced_istream = &infile;
+    scheduler_t scheduler;
+    if (scheduler.init(sched_inputs, NUM_OUTPUTS, sched_ops) !=
+        scheduler_t::STATUS_SUCCESS)
+        assert(false);
+    std::vector<std::string> sched_as_string =
+        run_lockstep_simulation(scheduler, NUM_OUTPUTS, TID_BASE);
+    for (int i = 0; i < NUM_OUTPUTS; i++) {
+        std::cerr << "cpu #" << i << " schedule: " << sched_as_string[i] << "\n";
+    }
+    assert(sched_as_string[0] == CORE0_SCHED_STRING);
+    assert(sched_as_string[1] == CORE1_SCHED_STRING);
+#endif
+}
+
+static void
+test_replay_as_traced_from_file(const char *testdir)
+{
+#if (defined(X86_64) || defined(ARM_64)) && defined(HAS_ZIP)
+    std::cerr << "\n----------------\nTesting replay as-traced from a file\n";
+    std::string path = std::string(testdir) + "/drmemtrace.threadsig.x64.tracedir";
+    std::string cpu_file =
+        std::string(testdir) + "/drmemtrace.threadsig.x64.tracedir/cpu_schedule.bin.zip";
+    // This checked-in trace has 8 threads on 9 cores, with one core starting
+    // out in a wait state.
+    static const char *const SCHED_STRING =
+        "Core #0: 3018553 \nCore #1: 3018556 \nCore #2: 3018557 \nCore #3: 3018558 "
+        "\nCore #4: 3018520 \nCore #5: 3018520 \nCore #6: 3018554 \nCore #7: 3018555 "
```

Comment:
I was also surprised it didn't have more.  There is a tradeoff here, though, where we want our GA CI to not take forever to run and we don't want to check in large binary files into our repository so we mostly run small apps.  This is the "threadsig" app we use for drmemtrace_samples.  This already stressed several corner cases (core 5 having to wait for core 4 right up front; merging duplicate entries) so it is a pretty good test; plus, the unit test has migration to test that aspect.  I could try to trace the threadsig app a few more times and hope to get more migration but it may not happen on this machine (plus it is work to then update the other tests that use this checked-in trace).  We also have a bigger internal proprietary test which does have a ton of migration.  I'm leaning toward just adding a comment here acknowledge the few migrations but pointed at the synthetic test having more migration.

Reply:
Ack. That's fine.